Lab 4/src/FiniteAutomata.java

```diff
@@ -0,0 +1,65 @@
+import java.io.*;
+import java.util.*;
+
+public class FiniteAutomata {
+    public List<String> allStates;
+    public List<String> alphabet;
+    public String q0;
+    public List<String> finalStates;
+    public Map<Pair<String, String>, List<String>> transitions;
+    public Boolean deterministic = true;
+
+    public void readFile(String filename) {
+        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
+
+            allStates = Arrays.asList(br.readLine().split(", "));
+            alphabet = Arrays.asList(br.readLine().split(", "));
+            q0 = br.readLine().strip();
+            finalStates = Arrays.asList(br.readLine().split(", "));
+
+            String line;
+            transitions = new HashMap<>();
+            while((line = br.readLine()) != null) {
+                String[] literals = line.split(",|=");
+                String fromState = literals[0].strip();
+                String symbol = literals[1].strip();
+                String toState = literals[2].strip();
+
+                Pair<String, String> transition = new Pair<>(fromState, symbol);
+                if (transitions.containsKey(transition)) {
+                    List<String> transitionStates = transitions.get(transition);
```

Comment:
You could hava had a Map<Pair<String, String>, String> instread of Map<Pair<String, String>, List<String>>, as you don't use the list anywhere if the FA is not a DFA